client/network/src/protocol/notifications/behaviour.rs

```diff
@@ -2070,3 +2073,1985 @@ impl NetworkBehaviour for Notifications {
+		assert!(std::matches!(notif.peers.get(&(peer, set_id)), Some(&PeerState::Disabled { .. })));
+
+		// remote opens a substream, verify that peer state is updated to `Incoming`
+		notif.inject_event(peer, conn, NotifsHandlerOut::OpenDesiredByRemote { protocol_index: 0 });
+
+		assert!(std::matches!(notif.peers.get(&(peer, set_id)), Some(&PeerState::Incoming { .. })));
+
+		notif.peerset_report_disconnect(peer, set_id);
+	}
+
+	#[test]
+	#[should_panic]
+	#[cfg(debug_assertions)]
+	fn peerset_report_accept_incoming_peer() {
+		let (mut notif, _peerset) = development_notifs();
+		let peer = PeerId::random();
+		let conn = ConnectionId::new(0usize);
+		let set_id = sc_peerset::SetId::from(0);
+		let connected = ConnectedPoint::Listener {
+			local_addr: Multiaddr::empty(),
+			send_back_addr: Multiaddr::empty(),
+		};
+
+		notif.inject_connection_established(&peer, &conn, &connected, None, 0usize);
+		assert!(std::matches!(notif.peers.get(&(peer, set_id)), Some(&PeerState::Disabled { .. })));
+
+		// remote opens a substream, verify that peer state is updated to `Incoming`
+		notif.inject_event(peer, conn, NotifsHandlerOut::OpenDesiredByRemote { protocol_index: 0 });
+		assert!(std::matches!(notif.peers.get(&(peer, set_id)), Some(&PeerState::Incoming { .. })));
+
+		assert!(std::matches!(
+			notif.incoming[0],
+			IncomingPeer { alive: true, incoming_id: sc_peerset::IncomingIndex(0), .. },
+		));
+
+		notif.peers.remove(&(peer, set_id));
+		notif.peerset_report_accept(sc_peerset::IncomingIndex(0));
+	}
+
+	#[test]
+	#[should_panic]
+	#[cfg(debug_assertions)]
+	fn peerset_report_accept_not_incoming_peer() {
+		let (mut notif, _peerset) = development_notifs();
+		let peer = PeerId::random();
+		let conn = ConnectionId::new(0usize);
+		let set_id = sc_peerset::SetId::from(0);
+		let connected = ConnectedPoint::Listener {
+			local_addr: Multiaddr::empty(),
+			send_back_addr: Multiaddr::empty(),
+		};
+		let mut conn_yielder = ConnectionYielder::new();
+
+		notif.inject_connection_established(&peer, &conn, &connected, None, 0usize);
+		assert!(std::matches!(notif.peers.get(&(peer, set_id)), Some(&PeerState::Disabled { .. })));
+
+		// remote opens a substream, verify that peer state is updated to `Incoming`
+		notif.inject_event(peer, conn, NotifsHandlerOut::OpenDesiredByRemote { protocol_index: 0 });
+		assert!(std::matches!(notif.peers.get(&(peer, set_id)), Some(&PeerState::Incoming { .. })));
+
+		assert!(std::matches!(
+			notif.incoming[0],
+			IncomingPeer { alive: true, incoming_id: sc_peerset::IncomingIndex(0), .. },
+		));
+
+		notif.peerset_report_connect(peer, set_id);
+		assert!(std::matches!(notif.peers.get(&(peer, set_id)), Some(&PeerState::Enabled { .. })));
+
+		let event = conn_yielder.open_substream(peer, 0, connected, vec![1, 2, 3, 4]);
+		notif.inject_event(peer, conn, event);
+
+		assert!(std::matches!(notif.peers.get(&(peer, set_id)), Some(&PeerState::Enabled { .. })));
+		notif.incoming[0].alive = true;
```

Comment:
Looks like inconsistent internal state. Is it what we are testing here?

Reply:
yes